File: src/main/java/es/danisales/datune/function/HarmonicFunction.java
package es.danisales.datune.function;

import es.danisales.datune.chords.chromatic.ChromaticChord;
import es.danisales.datune.chords.tonal.TonalChord;
import es.danisales.datune.tonality.ScaleRelativeDegreeException;
import es.danisales.datune.tonality.TonalityModern;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

public abstract class HarmonicFunction {
    private static FunctionCache functionCache = new FunctionCache();

    @Nullable public final ChromaticChord getChord(@NonNull TonalityModern tonality) {
        TonalChord tonalChord = TonalChord.from(tonality, this);
        ChromaticChord chromaticChord = (ChromaticChord)functionCache.get(tonalChord);
        if (functionCache.containsKey(tonalChord)) {
            chromaticChord = calculateChord(tonality);
            functionCache.put(tonalChord.clone(), ChromaticChord.immutableFrom(chromaticChord));
        }

        return chromaticChord;
    }
    @Nullable protected abstract ChromaticChord calculateChord(TonalityModern tonalityModern);
    @NonNull public abstract HarmonicFunction getShifted(int i);
}

File: src/function/HarmonicFunction.ts
import { DiatonicAltChord } from '../chords/diatonicalt/DiatonicAltChord';
import { TonalChord } from '../chords/tonal/TonalChord';
import { Tonality } from '../tonality/Tonality';

export abstract class HarmonicFunction {
    private static functionCache = new Map<string, DiatonicAltChord>();

    public getChord(tonality: Tonality): DiatonicAltChord | null {
        let tonalChord: TonalChord = TonalChord.from(tonality, this);
        let tonalChordHashCode = tonalChord.hashCode();
        let diatonicAltChord = HarmonicFunction.functionCache.get(tonalChordHashCode);
        if (diatonicAltChord == undefined) {
            diatonicAltChord = this.calculateChord(tonality);
            HarmonicFunction.functionCache.set(tonalChordHashCode, diatonicAltChord);
        }
        return diatonicAltChord;
    }

    protected abstract calculateChord(tonality: Tonality): DiatonicAltChord | null;
}